core/src/jvmMain/kotlin/org/aquila3d/core/vulkan/VkPhysicalDevice.kt

```diff
@@ -77,11 +80,79 @@ actual class VkPhysicalDevice(internal val device: org.lwjgl.vulkan.VkPhysicalDe
         return queueFamilies
     }
 
     actual fun getDeviceExtensions(): Map<String, Int> {
         return extensions
     }
+
+    actual fun getSwapchainFeatures(): SwapchainFeatures {
+        return swapFeatures
+    }
+
+    private fun getSurfaceCapabilities(): VkSurfaceCapabilitiesKHR {
+        val surfCaps = org.lwjgl.vulkan.VkSurfaceCapabilitiesKHR.calloc()
+        val err = vkGetPhysicalDeviceSurfaceCapabilitiesKHR(device, surface.surfaceHandle, surfCaps)
+        if (err != VK_SUCCESS) {
+            throw AssertionError("Failed to get physical device surface capabilities: ${VkResult(err)}")
+        }
+        val surfaceCapabilities = VkSurfaceCapabilitiesKHR(surfCaps)
+        surfCaps.free()
+        return surfaceCapabilities
+    }
+
+    private fun getSurfaceFormats(): List<VkSurfaceFormatKHR> {
+        val formatCountPointer = memAllocInt(1)
+        var err = vkGetPhysicalDeviceSurfaceFormatsKHR(device, surface.surfaceHandle, formatCountPointer, null)
+        val formatCount = formatCountPointer[0]
+        if (err != VK_SUCCESS) {
+            throw AssertionError("Failed to get number of physical device surface presentation modes: ${VkResult(err)}")
+        }
+        val formatsPointer = org.lwjgl.vulkan.VkSurfaceFormatKHR.calloc(formatCount)
+        err = vkGetPhysicalDeviceSurfaceFormatsKHR(device, surface.surfaceHandle, formatCountPointer, formatsPointer)
+        memFree(formatCountPointer)
+        if (err != VK_SUCCESS) {
+            throw AssertionError(
+                "Failed to get physical device surface presentation modes: ${VkResult(err)}")
+        }
+        val formats = mutableListOf<VkSurfaceFormatKHR>()
+        for (i in 0 until formatCount) {
+            val format = formatsPointer.get(i)
+            formats.add(VkSurfaceFormatKHR(format))
+        }
+        formatsPointer.free()
+        return formats
+    }
+
+    private fun getPresentationModes(): List<VkPresentModeKHR> {
+        val presentModeCountPointer = memAllocInt(1)
+        var err = vkGetPhysicalDeviceSurfacePresentModesKHR(device, surface.surfaceHandle, presentModeCountPointer, null)
+        val presentModeCount = presentModeCountPointer[0]
+        if (err != VK_SUCCESS) {
+            throw AssertionError("Failed to get number of physical device surface presentation modes: ${VkResult(err)}")
+        }
+        val pPresentModes = memAllocInt(presentModeCount)
+        err = vkGetPhysicalDeviceSurfacePresentModesKHR(device, surface.surfaceHandle, presentModeCountPointer, pPresentModes)
+        memFree(presentModeCountPointer)
+        if (err != VK_SUCCESS) {
+            throw AssertionError(
+                "Failed to get physical device surface presentation modes: ${VkResult(err)}")
+        }
+        val modes = mutableListOf<VkPresentModeKHR>()
+        for (i in 0 until presentModeCount) {
+            modes.add(VkPresentModeKHR.from(pPresentModes.get(i))!!)
```

Comment:
Instead of wrapping calls with force non null, why not change the `from` contract to return the value or throw an exception which in this case would have essentially the same result of an exception if the value ends up being null.

Reply:
Do you have a suggestion for the related case of:

```kotlin
queueFamilyIndicies.put(0, indices[VkQueueFamilies.VK_QUEUE_GRAPHICS]!!)
queueFamilyIndicies.put(1, indices[VkQueueFamilies.VK_QUEUE_PRESENTATION]!!)
```
In DefaultSwapchainCreator, around line 114?